### DISCUSSION TITLE
What kernel(s) allow coupling to a first order derivative?  Or square of first order derivative?

### DISCUSSION THREAD

**Discussion by EdSheltonUKAEA** [idaholab/moose#23554] [Q&A General] [answered]
I'm looking for an AD kernel which does the equivalent of coupling a time derivative to diffusion, but for the first derivative wrt x rather than the second derivative.

dy/dt = A dy/dx

rather than

dy/dt = D d^2y/dx^2

It looks like PrimaryConvetion and Convected Mesh have the right sort of idea, but are too specific to particular applications.  
https://mooseframework.inl.gov/source/kernels/PrimaryConvection.html
https://mooseframework.inl.gov/source/kernels/ConvectedMesh.html

I'm looking for a nice simple generic kernel which but can't see anything obvious in the names.  Also I need to be able to specify the coefficient A as a material property, and for the kernel to work with AD.

Does a kernel to meet my needs exist?

-----------------------------------------------------
P.S.  What would be even better is a kernel for:
dy/dt = A [dy/dx]^2

MatGradSquareCoupled looks like it could do this, but again maybe specific to a particular application and doesn't seem to have an AD version.
https://mooseframework.inl.gov/source/kernels/MatGradSquareCoupled.html

**Comment by GiudGiud** [comment, marked as answer]
Hello

MatGradSquareCoupled seems to do exactly what you want except it s not AD.
You can mix AD and non-AD kernels, the only issue is the AD-status of the material property. This kernel will only accept a non-AD one, and will retrieve the derivatives of the material properties to help form the Jacobian.

So you can either:
- rewrite that kernel with AD. This should be fairly simple. 
- use that kernel, duplicate your AD material property with a non-AD version and create the derivatives manually

Guillaume

**Reply by EdSheltonUKAEA** [reply]
Thank you!  This is really helpful!
When you say "create the derivatives manually" is this something that I have to do myself by doing hand calculations outside MOOSE and/or adding further lines of code for each derivative.  Or does MOOSE do these calculations itself just "manually" rather than using the automatic differentiation option.
I'm assuming the second one, but just want to be sure.

**Reply by GiudGiud** [reply]
Hello

So yes when people are making regular material properties and need the derivatives, they usually need to do the hand calculation outside of MOOSE, then create a new material property with a name like "dD_dT" and use that as the property. 

UNLESS it s a simple parsed expression and in this case the DerivativeParsedMaterial can do that for you. 

Guillaume

**Reply by GiudGiud** [reply]
but tbh the easy option here is to copy paste the kernel and make it AD. Just need to some these two letters in a bunch of places
